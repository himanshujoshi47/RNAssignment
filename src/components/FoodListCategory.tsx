import React, {memo, useEffect, useState} from 'react';
import {
  View,
  StyleSheet,
  Image,
  Text,
  TouchableOpacity,
  LayoutAnimation,
  Platform,
  UIManager,
} from 'react-native';
import {Icon} from 'react-native-elements';
import {CategoryElement} from '../network/models/FoodListModel';
import R from '../res';
import FoodListSubcategory from './FoodListSubcategory';

interface FoodListCategoryProps {
  itemIndex: number;
  item: CategoryElement;
  willExpand: boolean;
  didSelectCategory: (index: number) => void;
}

const FoodListCategory: React.FC<FoodListCategoryProps> = ({
  itemIndex,
  item,
  willExpand,
  didSelectCategory,
}) => {
  const [isExpanded, setIsExpanded] = useState<boolean>(willExpand);

  useEffect(() => {
    if (Platform.OS === 'android') {
      UIManager.setLayoutAnimationEnabledExperimental &&
        UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(() => {
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
    setIsExpanded(willExpand);
  }, [willExpand]);

  function expandItem() {
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
    didSelectCategory?.(itemIndex);
    setIsExpanded(!isExpanded);
  }

  let category = item.category;
  return (
    <View style={localStyles.container}>
      <View style={localStyles.touchableContainer}>
        <TouchableOpacity
          testID={R.id.FOOD_LIST_CAT_BTN}
          activeOpacity={0.9}
          onPress={() => expandItem()}>
          <View style={localStyles.touchableChild}>
            <Image
              source={{uri: category.imagePath}}
              style={[
                localStyles.touchableImage,
                {backgroundColor: category.colorCode},
              ]}
            />
            <View style={localStyles.touchableTextContainer}>
              <Text
                testID={R.id.FOOD_LIST_CAT_TITLE}
                style={[
                  localStyles.touchableText,
                  {color: category.colorCode},
                ]}>
                {category.categoryName}
              </Text>
              {category.servingSize && (
                <Text
                  testID={R.id.FOOD_LIST_CAT_SERVING_SIZE}
                  style={localStyles.touchableText}>
                  {` (${category.servingSize})`}
                </Text>
              )}
            </View>
            <Icon
              name={isExpanded ? 'arrow-drop-up' : 'arrow-drop-down'}
              color={R.colors.stroke.black}
              size={R.dimensions.dimen_24}
            />
          </View>
        </TouchableOpacity>
        {isExpanded && (
          <FoodListSubcategory
            quote={category.quote}
            colorCode={category.colorCode}
            subcategories={category.subcategories}
          />
        )}
      </View>
      {isExpanded && category.protip.length > 0 && (
        <View style={localStyles.protipContainer}>
          <View style={localStyles.tipContainer}>
            <Text style={localStyles.tip}>{R.strings.TIP}</Text>
          </View>
          <Text style={localStyles.tip}>{category.protip}</Text>
        </View>
      )}
    </View>
  );
};

const localStyles = StyleSheet.create({
  container: {
    borderRadius: R.dimensions.dimen_8,
    overflow: 'hidden',
  },
  touchableContainer: {
    paddingHorizontal: R.dimensions.dimen_8,
    paddingTop: R.dimensions.dimen_8,
    backgroundColor: R.colors.background.white,
  },
  touchableChild: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: R.dimensions.dimen_8,
  },
  touchableImage: {
    height: R.dimensions.dimen_44,
    width: R.dimensions.dimen_44,
    borderRadius: R.dimensions.dimen_4,
  },
  touchableTextContainer: {
    flex: 1,
    marginHorizontal: R.dimensions.dimen_16,
    flexDirection: 'row',
    fontWeight: '500',
  },
  touchableText: {
    fontWeight: '600',
    fontSize: R.dimensions.dimen_16,
    color: R.colors.text.primary,
  },
  protipContainer: {
    marginTop: R.dimensions.dimen_16,
    padding: R.dimensions.dimen_16,
    backgroundColor: R.colors.background.seaGreen,
    borderRadius: R.dimensions.dimen_8,
  },
  tipContainer: {
    flexDirection: 'row',
    alignSelf: 'flex-start',
    paddingHorizontal: R.dimensions.dimen_12,
    backgroundColor: R.colors.background.blue,
    borderRadius: R.dimensions.dimen_8,
    marginBottom: R.dimensions.dimen_8,
  },
  tip: {
    color: R.colors.background.white,
  },
});

export const FoodListItemSeparator = () => {
  return <View style={{height: R.dimensions.dimen_24}} />;
};

export default memo(FoodListCategory);
