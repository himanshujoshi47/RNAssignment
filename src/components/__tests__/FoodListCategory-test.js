import React from 'react';
import renderer, {act} from 'react-test-renderer';
import R from '../../res';
import FoodListCategory from '../FoodListCategory';

jest.useFakeTimers();

describe('renders and behave correctly', () => {
  let props = {
    itemIndex: 0,
    item: {
      category: {
        subcategories: [
          {
            items: [
              'Beef (93/7 ground)',
              'Bison',
              'Filet Mignon (trim fat)',
              'Kangaroo',
              'Sirloin (ground)',
              'Sirloin Strip Steak',
              'Veal (Ground)',
              'Venison',
            ],
            subCategoryname: 'Beef',
          },
          {
            items: ['Lean Loin Chop', 'Lean Pork Loin'],
            subCategoryname: 'Pork',
          },
          {
            items: [
              'Chicken',
              'Cornish Game Hen',
              'Emu',
              'Ostrich',
              'Quail',
              'Turkey',
            ],
            subCategoryname: 'Poultry',
          },
        ],
        quote: '',
        protip: '',
        imagePath:
          'https://s3-us-west-2.amazonaws.com/viking.approvedfoodlistimages/protein%403x.png',
        localImagePath: 'protein',
        categoryName: 'Lean Protein',
        colorCode: '#F168A4',
        servingSize: '4 Oz Servings',
      },
    },
    willExpand: false,
    didSelectCategory: jest.fn(),
  };

  const foodListCategory = renderer.create(<FoodListCategory {...props} />);
  const foodListCategoryInst = foodListCategory.root;

  it('renders correctly', () => {
    expect(foodListCategory).toMatchSnapshot();
  });

  it('should have required elements', () => {
    expect(
      foodListCategoryInst.findByProps({testID: R.id.FOOD_LIST_CAT_BTN}),
    ).toBeTruthy();
    expect(
      foodListCategoryInst.findByProps({testID: R.id.FOOD_LIST_CAT_TITLE}),
    ).toBeTruthy();
    expect(
      foodListCategoryInst.findByProps({
        testID: R.id.FOOD_LIST_CAT_SERVING_SIZE,
      }),
    ).toBeTruthy();
  });

  it('expand/collapse category on select', () => {
    act(() => {
      foodListCategoryInst
        .findByProps({testID: R.id.FOOD_LIST_CAT_BTN})
        .props.onPress();
    });
    expect(props.didSelectCategory).toBeCalled();
  });
});
