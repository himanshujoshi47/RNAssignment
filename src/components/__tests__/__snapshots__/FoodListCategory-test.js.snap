// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and behave correctly renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 4,
        "paddingHorizontal": 8,
        "paddingTop": 8,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="approved-food-cat-btn"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://s3-us-west-2.amazonaws.com/viking.approvedfoodlistimages/protein%403x.png",
            }
          }
          style={
            Array [
              Object {
                "borderRadius": 4,
                "height": 44,
                "width": 44,
              },
              Object {
                "backgroundColor": "#F168A4",
              },
            ]
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "fontWeight": "500",
              "marginHorizontal": 16,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#222222",
                  "fontSize": 16,
                  "fontWeight": "600",
                },
                Object {
                  "color": "#F168A4",
                },
              ]
            }
            testID="food-list-cat-title"
          >
            Lean Protein
          </Text>
          <Text
            style={
              Object {
                "color": "#222222",
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
            testID="food-list-cat-serving-size"
          >
             (4 Oz Servings)
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
